### PR TITLE
Add support for identity field for email verification jobs and tickets

### PR DESCRIPTION
**Pull request by jimmyjames**
### Changes

Adds support for specifying the (optional) `identity` object when creating email verification tickets or jobs. This is needed when verifying emails when the primary is social, enterprise, or passwordless. Also needed to verify secondary identities.

### References

- https://auth0.com/docs/api/management/v2#!/Tickets/post_email_verification
- https://auth0.com/docs/api/management/v2#!/Jobs/post_verification_email

### Testing

In addition to unit tests, I've also verified the changes manually using a real tenant and users.

- [X] This change adds test coverage
- [ ] This change has been tested on the latest version of the platform/language or why not

### Checklist

- [X] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [X] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [X] All existing and new tests complete without errors

### CODE REVIEW

File: src/main/java/com/auth0/json/mgmt/EmailVerificationIdentity.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.io.Serializable;
+
+/**
+ * Represents the identity object that can be sent on requests to create an email verification ticket or job.
+ * Related to {@link com.auth0.client.mgmt.JobsEntity} and {@link com.auth0.client.mgmt.TicketsEntity}.
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class EmailVerificationIdentity implements Serializable {
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/json/mgmt/EmailVerificationIdentity.java:15]: is the serializable really needed?

File: src/main/java/com/auth0/json/mgmt/EmailVerificationIdentity.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        return this.provider;
+    }
+
+    /**
+     * Get the user ID.
+     *
+     * @return the User ID.
+     */
+    @JsonProperty("user_id")
+    public String getUserId() {
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/json/mgmt/EmailVerificationIdentity.java:49]: The main method you added in the API client already receives a user id. What is this one for? Should it be renamed?

**Review by lbalmaceda — COMMENTED** [review]
I've not checked the tests

File: src/main/java/com/auth0/json/mgmt/EmailVerificationIdentity.java (hunk 1 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/json/mgmt/EmailVerificationIdentity.java:15]: Good catch. No, I don't think it is, will confirm. Looks like we have a few models that do implement Serializable for some reason.

File: src/main/java/com/auth0/json/mgmt/EmailVerificationIdentity.java (hunk 2 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/json/mgmt/EmailVerificationIdentity.java:49]: The `identity` object is defined as:
```
"identity": {
    "provider": "google-oauth2",
    "user_id": "117029267717860381713"
}
```

The `user_id` of the identity is required. For example, in the case of verifying a secondary email address, the `user_id` of the `identity` will be different than the top-level `user_id` field. For example, from the [send email verification job API docs](https://auth0.com/docs/api/management/v2#!/Jobs/post_verification_email):

```
{
  "user_id": "google-oauth2|1234",
  "client_id": "",
  "identity": {
    "user_id": "5457edea1b8f22891a000004",
    "provider": "google-oauth2"
  }
}
```

File: src/main/java/com/auth0/json/mgmt/EmailVerificationIdentity.java (hunk 1 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/json/mgmt/EmailVerificationIdentity.java]: Updated to not be Serializable.

File: src/test/java/com/auth0/client/mgmt/JobsEntityTest.java [hunk 3]
```diff
@@ -211,6 +213,55 @@ public void shouldThrowOnSendUserVerificationEmailWithNullUserId() {
...
+        exception.expect(IllegalArgumentException.class);
+        exception.expectMessage("'identity provider' cannot be null!");
+        api.jobs().sendVerificationEmail("google-oauth2|1234", null, new EmailVerificationIdentity(null, "user-id"));
+    }
+
+    @Test
+    public void shouldThrowOnSendUserVerificationEmailWithNullIdentityUserId() {
+        exception.expect(IllegalArgumentException.class);
+        // don't be too specific in testing the order we null-check provider and user-id
+        exception.expectMessage("'identity ");
```
Comment (lbalmaceda) [inline src/test/java/com/auth0/client/mgmt/JobsEntityTest.java:254]: looking at the code, the line that should trigger this failure is 
```
Asserts.assertNotNull(emailVerificationIdentity.getUserId(), "identity user id");
```
I'd expect the message check to be more specific. You mention in your comment you don't want to, what is the reason to ignore that?

File: src/test/java/com/auth0/client/mgmt/JobsEntityTest.java [hunk 4]
```diff
@@ -211,6 +213,55 @@ public void shouldThrowOnSendUserVerificationEmailWithNullUserId() {
...
+        api.jobs().sendVerificationEmail("google-oauth2|1234", null, new EmailVerificationIdentity("google-oauth2", null));
+    }
+
+    @Test
+    public void shouldThrowOnSendUserVerificationEmailWithNullIdentityProviderAndUserId() {
+        exception.expect(IllegalArgumentException.class);
+        exception.expectMessage("'identity provider' cannot be null!");
+        api.jobs().sendVerificationEmail("google-oauth2|1234", null, new EmailVerificationIdentity(null, null));
```
Comment (lbalmaceda) [inline src/test/java/com/auth0/client/mgmt/JobsEntityTest.java:262]: I don't think this use case is worth testing. But maybe yes the one where the last param, the email verification identity, is null.

**Review by lbalmaceda — CHANGES_REQUESTED** [review]

File: src/test/java/com/auth0/client/mgmt/JobsEntityTest.java (hunk 3 above)
Comment (jimmyjames) [reply src/test/java/com/auth0/client/mgmt/JobsEntityTest.java:254]: Good question. Because the assert will throw a RuntimeException, the order we check the params determines which error message you receive. It's a question of how tightly coupled we want the test to be to the implementation details (which param we check first) when both are null. If we think that the order of the checks defines the contract of the API and should never change, we should verify that in the tests. I tend to think that order is not part of the API (they both throw the same type of exception), so didn't want to couple the test to that implementation detail.

File: src/main/java/com/auth0/client/mgmt/JobsEntity.java [hunk 5]
```diff
@@ -61,6 +62,20 @@
...
+     * See https://auth0.com/docs/api/management/v2#!/Jobs/post_verification_email.
+     *
+     * @param userId   The user_id of the user to whom the email will be sent.
+     * @param clientId The id of the client, if not provided the global one will be used.
+     * @param emailVerificationIdentity The identity of the user. Required to verify primary identities when using social, enterprise, or passwordless connections. It is also required to verify secondary identities.
+     *
+     * @return a Request to execute.
+     */
+    public Request<Job> sendVerificationEmail(String userId, String clientId, EmailVerificationIdentity emailVerificationIdentity) {
         Asserts.assertNotNull(userId, "user id");
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/client/mgmt/JobsEntity.java:79]: Maybe instead we should add docs to the overloaded method that does not accept an `EmailVerificationIdentity`?

File: src/test/java/com/auth0/client/mgmt/JobsEntityTest.java (hunk 3 above)
Comment (jimmyjames) [reply src/test/java/com/auth0/client/mgmt/JobsEntityTest.java:254]: Ah, actually this check should be in the other test, `shouldThrowOnSendUserVerificationEmailWithNullIdentityProviderAndUserId` 😄. This one should indeed check the specific message. 

Reasoning still stands (willing to discuss further), just on the other test ;)

File: src/test/java/com/auth0/client/mgmt/JobsEntityTest.java (hunk 4 above)
Comment (jimmyjames) [reply src/test/java/com/auth0/client/mgmt/JobsEntityTest.java:262]: See [comment above](https://github.com/auth0/auth0-java/pull/293/files#r495041415) about what this test should be checking. Alternatively, we could remove this test as you suggest, it may indeed be too much of an edge case

File: src/main/java/com/auth0/client/mgmt/JobsEntity.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/client/mgmt/JobsEntity.java:79]: Specifically what I mean is adding a `@see` to on each of the `sendVerificationEmail` methods, with a link to the other overloaded method. Just a little tip to make sure developers know there is another method which may be useful for their use case (I'd expect IDEs auto-complete to be more used for this, but still not a bad practice). For example:

```java
/**
     * ...
     * @see JobsEntity#sendVerificationEmail(String, String, EmailVerificationIdentity)
     */
    public Request<Job> sendVerificationEmail(String userId, String clientId) {...}

    /**
     * ...
     * @see JobsEntity#sendVerificationEmail(String, String)
     */
    public Request<Job> sendVerificationEmail(String userId, String clientId, EmailVerificationIdentity emailVerificationIdentity) {...}

```

File: src/test/java/com/auth0/client/mgmt/JobsEntityTest.java (hunk 4 above)
Comment (jimmyjames) [reply src/test/java/com/auth0/client/mgmt/JobsEntityTest.java:254]: Per conversation, I removed this test and added a test for null email verification object: https://github.com/auth0/auth0-java/pull/293/commits/64df560e62fb13e290fb5951822a1629ec41a891

File: src/main/java/com/auth0/client/mgmt/JobsEntity.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lbalmaceda) [reply src/main/java/com/auth0/client/mgmt/JobsEntity.java:79]: That works for me 💯

**Review by lbalmaceda — APPROVED** [review]
